### ISSUE TITLE
CloudS3 storage backend: corrupted downloads

### ISSUE THREAD

**Issue by juhassi** [filesender/filesender#1417]
Hi,

I've been testing CloudS3 storage backend with Filesender 2.39 using private S3 storage (ceph / radosgw). To get better performance (and make our S3 happier), I've tried to increase the upload_chunk_size parameter from it's default of 5 MB.

I've noticed that if upload_chunk_size is increased, CloudS3 storage backend corrupts downloads.

I made a test file with individually numbered lines (1 .. a lot). Uploading this file to private S3 bucket works and if I grab the parts directly from the bucket and cat them together in right order, this file is intact. When downloading it via Filesender, file is the right length, but the 000000000000000000000000 part of the file gets repeated many times over.

After setting `$config['download_chunk_size'] = $config['upload_chunk_size'];`, this seems to be working better - but I haven't done enough tests (or found the actual bug in filesender's code) so I'm not sure if this a reliable fix.

Relevant(?) configuration is:

```
$config['max_transfer_size'] = '53687091200'; // 50 GB
$config['terasender_enabled'] = true;
$config['terasender_worker_count'] = 5;
$config['storage_type'] = 'cloudS3';
$config['cloud_s3_bucket'] = "";
$config['cloud_s3_version'] = "latest";
$config['cloud_s3_region'] = "";
$config['cloud_s3_endpoint'] = "https://private.cloud";
$config['cloud_s3_use_path_style_endpoint'] = false;
$config['cloud_s3_key'] = "yes";
$config['cloud_s3_secret'] = "yes";
$config['upload_chunk_size'] = 1024 * 1024 * 40;
$config['upload_crypted_chunk_size'] = $config['upload_chunk_size'] + 32;
$config['download_chunk_size'] = $config['upload_chunk_size'];
```

**Comment by monkeyiq** [comment]
I was at first thinking that server side archive generation might not be working with S3. It is strange that the zip works ok but the tar doesn't. Certainly a good piece of information to know to try to find a fix. Most of my testing was done against a fake local S3 server. IIRC I have some real S3 credentials I can use for testing. It might be useful to do a trace on a local S3 server again to try to isolate the difference (zip to tar).

**Comment by monkeyiq** [comment]
There is a check for upload_chunk_size and download_chunk_size but its conditions have not been updated when there were multiple cloud back ends added. I'll fix that now to make things more explicit for everyone.

https://github.com/filesender/filesender/blob/03bc5a8744ecf6f11411e812c7ee3196ebeead85/classes/utils/Config.class.php#L245

**Comment by monkeyiq** [comment]
Was the tar file for an encrypted or unencrypted transfer? 

These two use different code paths (deencryption is browser based so the tar has to be made in the browser).

**Comment by monkeyiq** [comment]
I am seeing this issue on unencrypted files to tar. I have not retested encrypted files to tar yet.

**Comment by juhassi** [comment]
Hi!

In my tests, .tar was created from an unencrypted download. Creating a .tar from encrypted download doesn't seem to be supported with my configuration (there is only a .zip button). Creating a .zip from encrypted download works.


I've applied the fix #1436 to my testing installation of 2.39 and can confirm it works for files smaller than download_chunk_size. But: if a file to be included in the tar is bigger than download_chunk_size, the tar is corrupted and only first download_chunk_size bytes are included in the archive.

When uploading this:
-rw-r--r--  1 user  group  563888783 Apr  4 09:44 fstest.txt

And then downloading it as a .tar, the archive is:
-rw-r--r-- 1 user  group  41943552 Apr 17 08:55 transfer_19_files_43f36048.tar

My chunk_size is 1024*1024*40.
